### PR TITLE
Added constructors with name argument to the target types

### LINKED ISSUE #1297: Create constructors with name for every target
For easier API usage.

~ 38 targets

- [ ] Add Constructor with `string name`. Call `this()`, after the default ctor
- [ ] Add default constructor if it isn't there
- [ ] write unit test with reflection to test all constructors (default + named). test also that default+named ctor will yield the same results.

NB: no extra contructor needed in base class (Target) because this will lead to (more) code duplication.

Before:

```c#
/// <summary>
/// Initializes a new instance of the <see cref="FileTarget" /> class.
/// </summary>
/// <remarks>
/// The default value of the layout is: <code>${longdate}|${level:uppercase=true}|${logger}|${message}</code>
/// </remarks>
public FileTarget()
{
	this.ArchiveNumbering = ArchiveNumberingMode.Sequence;
	this.maxArchiveFiles = 0;
	this.ConcurrentWriteAttemptDelay = 1;
	....
}  
```

after
```c#
/// <summary>
/// Initializes a new instance of the <see cref="FileTarget" /> class.
/// </summary>
/// <remarks>
/// The default value of the layout is: <code>${longdate}|${level:uppercase=true}|${logger}|${message}</code>
/// </remarks>
public FileTarget()
{
	this.ArchiveNumbering = ArchiveNumberingMode.Sequence;
	this.maxArchiveFiles = 0;
	this.ConcurrentWriteAttemptDelay = 1;
      ....
}
/// <summary>
/// Initializes a new instance of the <see cref="FileTarget" /> class.
/// </summary>
/// <param name="name">Name of the target.</param>
/// <remarks>
/// The default value of the layout is: <code>${longdate}|${level:uppercase=true}|${logger}|${message}</code>
/// </remarks>
public FileTarget(string name) : this()
{
	this.Name = name;
}    
```

### PR DESCRIPTION
**Pull request by 304NotModified**
Added "name" to the constructors of all the Targets.

- For regular targets : added constructor `target(string name)`
- For Wrapper targets:  added constructor `target(string name, params Target target)`
- For Wrapper targets:  added constructor `target(string name, params Target[] targetst)`

supersedes https://github.com/NLog/NLog/pull/1370

fixes #1297

Thanks @flyingcroissant !

### CODE REVIEW

File: src/NLog/Targets/Wrappers/AsyncTargetWrapper.cs [hunk 1]
```diff
@@ -87,13 +87,23 @@ public class AsyncTargetWrapper : WrapperTargetBase
         /// Initializes a new instance of the <see cref="AsyncTargetWrapper" /> class.
         /// </summary>
         public AsyncTargetWrapper()
-            : this(null)
+            : this((Target)null)
```
Comment (304NotModified) [inline src/NLog/Targets/Wrappers/AsyncTargetWrapper.cs]: Breaking?

File: src/NLog/Targets/Wrappers/RoundRobinGroupTarget.cs [hunk 2]
```diff
@@ -74,6 +74,16 @@ public RoundRobinGroupTarget()
         /// <summary>
         /// Initializes a new instance of the <see cref="RoundRobinGroupTarget" /> class.
         /// </summary>
+        /// <param name="name">Name of the target.</param>
+        public RoundRobinGroupTarget(string name)
```
Comment (304NotModified) [inline src/NLog/Targets/Wrappers/RoundRobinGroupTarget.cs]: Need string +params?

**Comment by 304NotModified** [comment]
Todo wrapper targets needs another approach

File: src/NLog/Targets/FileTarget.cs [hunk 3]
```diff
@@ -240,16 +252,24 @@ public Layout FileName
...
-            if (simpleLayout != null && simpleLayout.IsFixedText)
-            {
-                cachedCleanedFileNamed = CleanupInvalidFileNameChars(simpleLayout.FixedText);
-            }
-            else
-            {
-                //clear cache
-                cachedCleanedFileNamed = null;
-            }
+                var simpleLayout = value as SimpleLayout;
```
Comment (304NotModified) [inline src/NLog/Targets/FileTarget.cs]: Fix indentation

**Comment by bhaeussermann** [comment]
AppVeyor/pr passed, so I assume it is safe to merge this.

**Comment by 304NotModified** [comment]
:+1:

**Pull request merged**